### ISSUE TITLE
`bundle install --deployment` failing after `bundle exec`

### ISSUE THREAD

**Issue by jabenninghoff** [bundler/bundler#5346]
I use Travis CI to validate my github-pages website. Recently, my Travis builds started failing with the following error:

```
You are trying to install in deployment mode after changing
your Gemfile. Run `bundle install` elsewhere and add the
updated Gemfile.lock to version control.

You have deleted from the Gemfile:
* json
```

In my Gemfile, I dynamically load the github-pages version:

```ruby
source 'https://rubygems.org'

require 'json'
require 'open-uri'
version = JSON.parse(open('https://pages.github.com/versions.json').read)

ruby version['ruby']
gem 'github-pages', version['github-pages'], group: :jekyll_plugins
gem 'html-proofer', group: :htmlproofer
```

After some testing, I discovered that `bundle install` generated a `Gemfile.lock` without `json` listed in the `DEPENDENCIES` section. Running `bundle exec` added `json` to `DEPENDENCIES`:

```diff
--- Gemfile.lock.install	2017-01-22 08:38:39.000000000 -0600
+++ Gemfile.lock.exec	2017-01-22 08:38:57.000000000 -0600
@@ -209,6 +209,7 @@
 DEPENDENCIES
   github-pages (= 116)
   html-proofer
+  json
 
 RUBY VERSION
    ruby 2.3.3p222
```

I found that as a workaround I could add json as a required gem, after which json was always listed in `DEPENDENCIES`:

```ruby
source 'https://rubygems.org'

require 'json'
require 'open-uri'
version = JSON.parse(open('https://pages.github.com/versions.json').read)

ruby version['ruby']
gem 'github-pages', version['github-pages'], group: :jekyll_plugins
gem 'html-proofer', group: :htmlproofer
gem 'json', group: :travis_ci # requiring json a second time fixes Travis CI builds
```

Testing further, I discovered that this behavior changed between 1.13.7 and 1.14.1; my Travis builds started failing after Travis upgraded to 1.14.1 on their build server. I've written a [repro script](https://gist.github.com/jabenninghoff/a1571ff097f65bade410d7f353d8ceff) that works with 1.13.7 but fails on 1.14.1.

**Comment by segiddins** [comment]
While this might look like a regression, it's actually a funny coincidence this ever worked at all.

Internally, RubyGems patches `require` to try and call `Kernel#gem` if it needs to activate a gem to require the given file (as is the case with JSON). However, it calls `gem` on `self`, rather than explicitly on `Kernel`. Since the Gemfile DSL implements `gem`... https://github.com/rubygems/rubygems/pull/1822 would fix this in RubyGems, but I'm trying to understand why a change in behavior happened at all. In the meantime, explicitly calling `Kernel.require` instead of `require` might be an acceptable workaround.

**Comment by segiddins** [comment]
The issue was introduced because as of 1.14 the feature flag for plugins defaults to true, and plugins cause the Gemfile to be evaluated a second time. @asutoshpalai any idea if we can avoid that?

**Comment by jabenninghoff** [comment]
Thanks for the quick response! I changed my Gemfile to explicitly call `Kernel.require` and it fixes the problem.

```ruby
source 'https://rubygems.org'

# use Kernel.require as a workaround per https://github.com/bundler/bundler/issues/5346
Kernel.require 'json'
Kernel.require 'open-uri'
version = JSON.parse(open('https://pages.github.com/versions.json').read)

ruby version['ruby']
gem 'github-pages', version['github-pages'], group: :jekyll_plugins
gem 'html-proofer', group: :htmlproofer
```

**Comment by segiddins** [comment]
We could also add something like the following to bundler:

```diff
diff --git a/lib/bundler/dsl.rb b/lib/bundler/dsl.rb
index cdbae076f..9ddc2b3c7 100644
--- a/lib/bundler/dsl.rb
+++ b/lib/bundler/dsl.rb
@@ -86,6 +86,7 @@ module Bundler
     end
 
     def gem(name, *args)
+      return super if caller(1 ,1).first.include?("rubygems/core_ext/kernel_require.rb")
       options = args.last.is_a?(Hash) ? args.pop.dup : {}
       version = args || [">= 0"]
 
```

**Comment by segiddins** [comment]
Issue fixed in RubyGems 2.6.10 as well.
